pkg/cloud/azure/actuators/machine/actuator_test.go

```diff
@@ -748,3 +750,91 @@ func TestMachineEvents(t *testing.T) {
 		})
 	}
 }
+
+func TestStatusCodeBasedCreationErrors(t *testing.T) {
+	machine, err := stubMachine()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cases := []struct {
+		name       string
+		operation  func(actuator *Actuator, machine *machinev1.Machine)
+		event      string
+		statusCode uint
+		requeable  bool
+	}{
+		{
+			name:       "InvalidConfig",
+			event:      "Warning FailedCreate InvalidConfiguration: failed to reconcile machine \"azure-actuator-testing-machine\": compute.VirtualMachinesClient#CreateOrUpdate: MOCK: StatusCode=400",
+			statusCode: 400,
+			requeable:  false,
+		},
+		{
+			name:       "CreateMachine",
+			event:      "Warning FailedCreate CreateError: failed to reconcile machine \"azure-actuator-testing-machine\"s: failed to create vm azure-actuator-testing-machine : failed to create or get machine: failed to create or get machine: compute.VirtualMachinesClient#CreateOrUpdate: MOCK: StatusCode=300",
+			statusCode: 300,
+			requeable:  true,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			cs := controllerfake.NewFakeClient(stubAzureCredentialsSecret(), &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "azure-actuator-user-data-secret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"userData": []byte("S3CR3T"),
+				},
+			})
+
+			mockCtrl := gomock.NewController(t)
+			networkSvc := mock_azure.NewMockService(mockCtrl)
+			vmSvc := mock_azure.NewMockService(mockCtrl)
+
+			eventsChannel := make(chan string, 1)
+
+			machineActuator := NewActuator(ActuatorParams{
+				Client:     fake.NewSimpleClientset(machine).MachineV1beta1(),
+				CoreClient: cs,
+				ReconcilerBuilder: func(scope *actuators.MachineScope) *Reconciler {
+					return &Reconciler{
+						scope:                scope,
+						networkInterfacesSvc: networkSvc,
+						virtualMachinesSvc:   vmSvc,
+					}
+				},
+				// use fake recorder and store an event into one item long buffer for subsequent check
+				EventRecorder: &record.FakeRecorder{
+					Events: eventsChannel,
+				},
+			})
+
+			networkSvc.EXPECT().CreateOrUpdate(gomock.Any(), gomock.Any()).Return(nil).Times(1)
+			azureErr := autorest.NewError("compute.VirtualMachinesClient", "CreateOrUpdate", "MOCK")
+			azureErr.StatusCode = tc.statusCode
+			wrapErr := errors.Wrapf(azureErr, "failed to create or get machine")
+			vmSvc.EXPECT().CreateOrUpdate(gomock.Any(), gomock.Any()).Return(wrapErr).Times(1)
+			vmSvc.EXPECT().Get(gomock.Any(), gomock.Any()).Return(nil, autorest.NewError("compute.VirtualMachinesClient", "Get", "MOCK")).Times(1)
```

Comment:
why we need to mock this `get`?

Reply:
https://github.com/openshift/cluster-api-provider-azure/blob/c1d771c6c17a00b9c07f8b17484d170fe8dc0ed0/pkg/cloud/azure/actuators/machine/reconciler.go#L539

This first checks if the machine does not exist. Without it it will never reach the creation.